File: ss1-backend-node/src/modules/auth/auth.service.ts
import {
  Injectable,
  UnauthorizedException,
  ConflictException,
  Inject,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { randomInt } from 'crypto';
import { MailService } from '../mail/mailtrap.service';
import { UserModel } from '../users/entities/user.entity';
import { UsersService } from '../users/users.service';
import { UpdateProfileDto } from './dto/update-profile.dto';
import { PatientModel } from '../patients/entities/patient.entity';
import type { ModelClass } from 'objection';

type TwoFaPurpose = 'login' | 'enable' | 'disable';

interface TwoFaJwtPayload {
  sub: number;
  type: '2fa';
  purpose: TwoFaPurpose;
}

interface PartialUserUpdate {
  last_login_at?: Date | null;
  two_fa_secret?: string | null;
  two_fa_expires_at?: Date | null;
  two_fa_attempts?: number;
  two_fa_enabled?: boolean;
  password_reset_token?: string | null;
  password_reset_expires?: string | null;
  password_hash?: string;
}

@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwt: JwtService,
    private readonly mail: MailService,
  ) {}

  // --- LocalStrategy usa esto ---
  async validateUser(emailOrUsername: string, password: string) {
    const user = await this.usersService.findByEmailOrUsername(emailOrUsername);
    if (!user || !user.is_active) return null;

    const ok = await bcrypt.compare(password, user.password_hash ?? '');
    if (!ok) return null;

    return user;
  }

  // --- Access JWT ---
  async issueAccessToken(user: UserModel): Promise<string> {
    const payload = {
      sub: user.id,
      email: user.email,
      username: user.username,
      roleId: user.role_id,
    };

    // "expira en bastante tiempo": configúralo por env, ej: 7d, 30d, etc.
    return this.jwt.signAsync(payload);
  }

  publicUser(user: UserModel) {
    // Extraer códigos de permisos del rol
    const permissions = user.role?.permissions?.map((p) => p.code) ?? [];

    return {
      id: user.id,
      email: user.email,
      username: user.username,
      role_id: user.role_id,
      role_name: user.role?.name ?? null,
      role_label: user.role?.label ?? null,
      two_fa_enabled: user.two_fa_enabled,
      permissions,
      employee: user.employee
        ? {
            id: user.employee.id,
            first_name: user.employee.first_name,
            last_name: user.employee.last_name,
            license_number: user.employee.license_number,
            status: user.employee.status,
          }
        : null,
      patient: user.patient
        ? {
            id: user.patient.id,
            first_name: user.patient.first_name,
            last_name: user.patient.last_name,
            dob: user.patient.dob,
            phone: user.patient.phone,
            email: user.patient.email,
            status: user.patient.status,
          }
        : null,
    };
  }

  async updateLastLogin(userId: number) {
    await this.usersService.update(userId, {
      last_login_at: new Date(),
    });
  }

  // --- 2FA: generar y enviar código ---
  private generateCode(): string {
    // 6 dígitos, con ceros a la izquierda
    const n = randomInt(0, 1_000_000);
    return n.toString().padStart(6, '0');
  }

  private async storeTwoFaCode(userId: number, code: string, ttlMinutes = 10) {
    const codeHash = await bcrypt.hash(code, 10);
    const expires = new Date(Date.now() + ttlMinutes * 60 * 1000);

    const updateData: PartialUserUpdate = {
      two_fa_secret: codeHash,
      two_fa_expires_at: expires,
      two_fa_attempts: 0,
    };

    await this.usersService.update(userId, updateData);
  }

  private async sendTwoFaEmail(
    to: string,
    code: string,
    purpose: TwoFaPurpose,
  ) {
    const subjectMap: Record<TwoFaPurpose, string> = {
      login: 'Tu código de verificación',
      enable: 'Código para activar 2FA',
      disable: 'Código para desactivar 2FA',
    };

    const subject = subjectMap[purpose];
    const text = `Tu código es: ${code}\n\nEste código expira en pocos minutos.`;

    await this.mail.sendMail({
      to,
      subject,
      text,
    });
  }

  // --- ChallengeId: JWT temporal (NO access token) ---
  private async issueTwoFaChallenge(
    userId: number,
    purpose: TwoFaPurpose,
  ): Promise<string> {
    return this.jwt.signAsync({ sub: userId, type: '2fa', purpose });
  }

  private verifyTwoFaChallenge(
    challengeId: string,
  ): { userId: number; purpose: TwoFaPurpose } | null {
    try {
      const payload = this.jwt.verify<TwoFaJwtPayload>(challengeId);
      if (!payload || payload.type !== '2fa') return null;
      return { userId: payload.sub, purpose: payload.purpose };
    } catch {
      return null;
    }
  }

  // --- Login flow con 2FA ---
  async startTwoFaLogin(userId: number): Promise<string> {
    const user = await this.usersService.findById(userId);
    if (!user || !user.is_active)
      throw new UnauthorizedException('Usuario inválido');
    const code = this.generateCode();

    await this.storeTwoFaCode(userId, code, 10);
    await this.sendTwoFaEmail(user.email, code, 'login');

    return this.issueTwoFaChallenge(userId, 'login');
  }

  async verifyTwoFaCode(
    challengeId: string,
    code: string,
  ): Promise<{ ok: true; user: UserModel } | { ok: false; reason: string }> {
    const challenge = this.verifyTwoFaChallenge(challengeId);
    if (!challenge || challenge.purpose !== 'login')
      return { ok: false, reason: 'Challenge inválido o expirado' };

    const user = await this.usersService.findById(challenge.userId);
    if (!user || !user.is_active)
      return { ok: false, reason: 'Usuario inválido' };

    // validar expiración
    if (!user.two_fa_expires_at)
      return { ok: false, reason: 'Código no solicitado' };
    const expiresAt = new Date(user.two_fa_expires_at).getTime();
    if (Date.now() > expiresAt) return { ok: false, reason: 'Código expirado' };

    // intentos (si agregas two_fa_attempts)
    const attempts = Number(user.two_fa_attempts ?? 0);
    if (attempts >= 5)
      return {
        ok: false,
        reason: 'Demasiados intentos. Solicita un nuevo código.',
      };

    const ok = await bcrypt.compare(code, user.two_fa_secret ?? '');
    if (!ok) {
      const updateData: PartialUserUpdate = {
        two_fa_attempts: attempts + 1,
      };
      await this.usersService.update(user.id, updateData);
      return { ok: false, reason: 'Código inválido' };
    }

    // consumir código (one-time)
    const updateData: PartialUserUpdate = {
      two_fa_secret: null,
      two_fa_expires_at: null,
      two_fa_attempts: 0,
    };
    await this.usersService.update(user.id, updateData);

    return { ok: true, user };
  }

  // --- Enable/Disable flows ---
  async startTwoFaToggle(userId: number, action: 'enable' | 'disable') {
    const user = await this.usersService.findById(userId);
    if (!user) throw new UnauthorizedException('Usuario inválido');
    const code = this.generateCode();

    await this.storeTwoFaCode(userId, code, 10);
    await this.sendTwoFaEmail(user.email, code, action);

    return this.issueTwoFaChallenge(userId, action);
  }

  async confirmTwoFaToggle(
    userId: number,
    challengeId: string,
    code: string,
    action: 'enable' | 'disable',
  ): Promise<{ ok: true } | { ok: false; reason: string }> {
    const challenge = this.verifyTwoFaChallenge(challengeId);
    if (
      !challenge ||
      challenge.userId !== userId ||
      challenge.purpose !== action
    ) {
      return { ok: false, reason: 'Challenge inválido o expirado' };
    }

    const user = await this.usersService.findById(userId);
    if (!user) return { ok: false, reason: 'Usuario inválido' };

    // Reutilizamos la misma validación que login:
    const expiresAt = user.two_fa_expires_at
      ? new Date(user.two_fa_expires_at).getTime()
      : 0;
    if (!expiresAt || Date.now() > expiresAt)
      return { ok: false, reason: 'Código expirado' };

    const attempts = Number(user.two_fa_attempts ?? 0);
    if (attempts >= 5)
      return {
        ok: false,
        reason: 'Demasiados intentos. Solicita un nuevo código.',
      };

    const ok = await bcrypt.compare(code, user.two_fa_secret ?? '');
    if (!ok) {
      const updateData: PartialUserUpdate = {
        two_fa_attempts: attempts + 1,
      };
      await this.usersService.update(user.id, updateData);
      return { ok: false, reason: 'Código inválido' };
    }

    // aplicar cambio y consumir código
    const updateData: PartialUserUpdate = {
      two_fa_enabled: action === 'enable',
      two_fa_secret: null,
      two_fa_expires_at: null,
      two_fa_attempts: 0,
    };
    await this.usersService.update(user.id, updateData);

    return { ok: true };
  }

  // --- Password Reset ---
  async requestPasswordReset(email: string): Promise<void> {
    const user = await this.usersService.findByEmailOrUsername(email);
    // Por seguridad, no revelamos si el email existe o no
    if (!user || !user.is_active) return;

    const code = this.generateCode();
    const codeHash = await bcrypt.hash(code, 10);
    const expires = new Date(Date.now() + 15 * 60 * 1000); // 15 minutos

    const updateData: PartialUserUpdate = {
      password_reset_token: codeHash,
      password_reset_expires: expires.toISOString(),
    };
    await this.usersService.update(user.id, updateData);

    await this.mail.sendMail({
      to: user.email,
      subject: 'Código de recuperación de contraseña',
      text: `Tu código de recuperación es: ${code}\n\nEste código expira en 15 minutos.`,
    });
  }

  async resetPassword(
    email: string,
    code: string,
    newPassword: string,
  ): Promise<{ ok: true } | { ok: false; reason: string }> {
    const user = await this.usersService.findByEmailOrUsername(email);
    if (!user || !user.is_active)
      return { ok: false, reason: 'Usuario inválido' };

    if (!user.password_reset_token || !user.password_reset_expires)
      return { ok: false, reason: 'Código no solicitado' };

    const expiresAt = new Date(user.password_reset_expires).getTime();
    if (Date.now() > expiresAt) return { ok: false, reason: 'Código expirado' };

    const ok = await bcrypt.compare(code, user.password_reset_token);
    if (!ok) return { ok: false, reason: 'Código inválido' };

    // Hash de la nueva contraseña
    const passwordHash = await bcrypt.hash(newPassword, 10);

    // Actualizar contraseña y limpiar token
    const updateData: PartialUserUpdate = {
      password_hash: passwordHash,
      password_reset_token: null,
      password_reset_expires: null,
    };
    await this.usersService.update(user.id, updateData);

    return { ok: true };
  }

  // --- Change Password (usuario autenticado) ---
  async changePassword(
    userId: number,
    currentPassword: string,
    newPassword: string,
  ): Promise<{ ok: true } | { ok: false; reason: string }> {
    const user = await this.usersService.findById(userId);
    if (!user || !user.is_active)
      return { ok: false, reason: 'Usuario inválido' };

    // Verificar que la contraseña actual sea correcta
    const isCurrentPasswordValid = await bcrypt.compare(
      currentPassword,
      user.password_hash ?? '',
    );
    if (!isCurrentPasswordValid)
      return { ok: false, reason: 'Contraseña actual incorrecta' };

    // Hash de la nueva contraseña
    const passwordHash = await bcrypt.hash(newPassword, 10);

    // Actualizar contraseña
    const updateData: PartialUserUpdate = {
      password_hash: passwordHash,
    };
    await this.usersService.update(user.id, updateData);

    return { ok: true };
  }

  // --- Update Profile (datos personales) ---
  async updateProfile(
    userId: number,
    dto: UpdateProfileDto,
  ): Promise<UserModel> {
    const user = await this.usersService.findById(userId);
    if (!user || !user.is_active) {
      throw new UnauthorizedException('Usuario inválido');
    }

    const trx = await PatientModel.startTransaction();

    try {
      // Actualizar username si se proporciona
      if (dto.username !== undefined) {
        // Verificar si el nuevo username ya existe
        const existingUsername = await this.usersService.findByEmailOrUsername(
          dto.username,
          trx,
        );
        if (existingUsername && existingUsername.id !== userId) {
          throw new ConflictException('El username ya está en uso');
        }

        await this.usersService.update(userId, { username: dto.username }, trx);
      }

      // Si el usuario es paciente, actualizar datos del paciente
      if (user.patient?.id) {
        const patientUpdate: Partial<PatientModel> = {};

        // Solo incluir los campos que se proporcionaron en el DTO
        if (dto.phone !== undefined) patientUpdate.phone = dto.phone;
        if (dto.email !== undefined) patientUpdate.email = dto.email;
        if (dto.address !== undefined) patientUpdate.address = dto.address;
        if (dto.gender !== undefined) patientUpdate.gender = dto.gender;
        if (dto.marital_status !== undefined)
          patientUpdate.marital_status = dto.marital_status;
        if (dto.occupation !== undefined)
          patientUpdate.occupation = dto.occupation;
        if (dto.education_level !== undefined)
          patientUpdate.education_level = dto.education_level;
        if (dto.emergency_contact_name !== undefined)
          patientUpdate.emergency_contact_name = dto.emergency_contact_name;
        if (dto.emergency_contact_relationship !== undefined)
          patientUpdate.emergency_contact_relationship =
            dto.emergency_contact_relationship;
        if (dto.emergency_contact_phone !== undefined)
          patientUpdate.emergency_contact_phone = dto.emergency_contact_phone;

        // Si hay campos de paciente para actualizar
        if (Object.keys(patientUpdate).length > 0) {
          await PatientModel.query(trx).patchAndFetchById(
            user.patient.id,
            patientUpdate,
          );
        }
      }

      await trx.commit();

      // Retornar el usuario actualizado con todas sus relaciones
      const updatedUser = await this.usersService.findById(userId);
      if (!updatedUser) {
        throw new UnauthorizedException(
          'Error al recuperar el usuario actualizado',
        );
      }
      return updatedUser;
    } catch (error) {
      await trx.rollback();
      throw error;
    }
  }
}

File: ss1-backend-python/app/services/auth/auth_service.py
from datetime import datetime, timedelta, timezone
from secrets import randbelow
from app.core.config import settings
from app.core.security import (
    verify_hash,
    hash_value,
    create_access_token,
    create_2fa_challenge,
    decode_2fa_challenge,
)
from app.db.repositories.users_repo import UsersRepo
from fastapi import HTTPException

MAX_2FA_ATTEMPTS = 5
TWOFA_TTL_MINUTES = 10

def _gen_6_digit_code() -> str:
    return str(randbelow(1_000_000)).zfill(6)

class AuthService:
    def __init__(self, users: UsersRepo, mail):
        self.users = users
        self.mail = mail

    async def authenticate_user(self, email_or_username: str, password: str):
        user = await self.users.find_by_email_or_username(email_or_username)
        if not user or not user.is_active:
            return None
        if not verify_hash(password, user.password_hash):
            return None
        return user

    async def issue_access_token(self, user) -> str:
        return create_access_token(
            subject=str(user.id),
            secret=settings.JWT_ACCESS_SECRET,
            expires_in=settings.JWT_ACCESS_EXPIRES_IN,
            extra={"email": user.email, "username": user.username, "role_id": user.role_id},
        )

    def public_user(self, user):
        # Extraer códigos de permisos del rol
        permissions = [p.code for p in user.role.permissions] if user.role and user.role.permissions else []

        return {
            "id": user.id,
            "email": user.email,
            "username": user.username,
            "role_id": user.role_id,
            "role_name": user.role.name if user.role else None,
            "role_label": user.role.label if user.role else None,
            "two_fa_enabled": user.two_fa_enabled,
            "permissions": permissions,
            "employee": user.employee if user.employee else None,
            "patient": user.patient if user.patient else None,
        }

    async def update_last_login(self, user_id: int):
        await self.users.patch_user(user_id, {"last_login_at": datetime.now(timezone.utc)})

    async def _store_twofa_code(self, user_id: int, code: str):
        expires_at = datetime.now(timezone.utc) + timedelta(minutes=TWOFA_TTL_MINUTES)
        await self.users.patch_user(
            user_id,
            {
                "two_fa_secret": hash_value(code),
                "two_fa_expires_at": expires_at,
                "two_fa_attempts": 0,
            },
        )

    async def _send_twofa_email(self, to: str, code: str, purpose: str):
        subject_map = {
            "login": "Tu código de verificación",
            "enable": "Código para activar 2FA",
            "disable": "Código para desactivar 2FA",
        }
        subject = subject_map.get(purpose, "Código de verificación")
        text = f"Tu código es: {code}\n\nEste código expira en pocos minutos."
        self.mail.send_text_email(to=to, subject=subject, text=text)

    async def start_twofa(self, user_id: int, purpose: str) -> str:
        user = await self.users.find_by_id(user_id)
        code = _gen_6_digit_code()
        await self._store_twofa_code(user_id, code)
        await self._send_twofa_email(user.email, code, purpose)

        return create_2fa_challenge(
            user_id=user_id,
            purpose=purpose,  # login/enable/disable
            secret=settings.JWT_2FA_SECRET,
            expires_in=settings.JWT_2FA_EXPIRES_IN,
        )

    async def verify_twofa_login(self, challenge_id: str, code: str):
        payload = decode_2fa_challenge(challenge_id, settings.JWT_2FA_SECRET)
        if not payload or payload.get("purpose") != "login":
            return {"ok": False, "reason": "Challenge inválido o expirado"}

        user_id = int(payload["sub"])
        user = await self.users.find_by_id(user_id)
        if not user or not user.is_active:
            return {"ok": False, "reason": "Usuario inválido"}

        if not user.two_fa_expires_at:
            return {"ok": False, "reason": "Código no solicitado"}
        if datetime.now(timezone.utc) > user.two_fa_expires_at:
            return {"ok": False, "reason": "Código expirado"}

        attempts = int(user.two_fa_attempts or 0)
        if attempts >= MAX_2FA_ATTEMPTS:
            return {"ok": False, "reason": "Demasiados intentos. Solicita un nuevo código."}

        if not verify_hash(code, user.two_fa_secret or ""):
            await self.users.patch_user(user.id, {"two_fa_attempts": attempts + 1})
            return {"ok": False, "reason": "Código inválido"}

        # consumir código
        await self.users.patch_user(
            user.id,
            {"two_fa_secret": None, "two_fa_expires_at": None, "two_fa_attempts": 0},
        )
        return {"ok": True, "user": user}

    async def confirm_twofa_toggle(self, user_id: int, challenge_id: str, code: str, action: str):
        payload = decode_2fa_challenge(challenge_id, settings.JWT_2FA_SECRET)
        if not payload or payload.get("purpose") != action:
            return {"ok": False, "reason": "Challenge inválido o expirado"}

        if int(payload["sub"]) != user_id:
            return {"ok": False, "reason": "Challenge no corresponde al usuario"}

        user = await self.users.find_by_id(user_id)
        if not user:
            return {"ok": False, "reason": "Usuario inválido"}

        if not user.two_fa_expires_at or datetime.now(timezone.utc) > user.two_fa_expires_at:
            return {"ok": False, "reason": "Código expirado"}

        attempts = int(user.two_fa_attempts or 0)
        if attempts >= MAX_2FA_ATTEMPTS:
            return {"ok": False, "reason": "Demasiados intentos. Solicita un nuevo código."}

        if not verify_hash(code, user.two_fa_secret or ""):
            await self.users.patch_user(user.id, {"two_fa_attempts": attempts + 1})
            return {"ok": False, "reason": "Código inválido"}

        # aplicar cambio + consumir código
        await self.users.patch_user(
            user.id,
            {
                "two_fa_enabled": (action == "enable"),
                "two_fa_secret": None,
                "two_fa_expires_at": None,
                "two_fa_attempts": 0,
            },
        )
        return {"ok": True}

    async def request_password_reset(self, email: str):
        """Solicita recuperación de contraseña - envía código al correo"""
        user = await self.users.find_by_email_or_username(email)
        # Por seguridad, no revelamos si el email existe o no
        if not user or not user.is_active:
            return

        code = _gen_6_digit_code()
        expires_at = datetime.now(timezone.utc) + timedelta(minutes=15)

        await self.users.patch_user(
            user.id,
            {
                "password_reset_token": hash_value(code),
                "password_reset_expires": expires_at,
            },
        )

        self.mail.send_text_email(
            to=user.email,
            subject="Código de recuperación de contraseña",
            text=f"Tu código de recuperación es: {code}\n\nEste código expira en 15 minutos.",
        )

    async def reset_password(self, email: str, code: str, new_password: str):
        """Resetea la contraseña usando el código enviado"""
        user = await self.users.find_by_email_or_username(email)
        if not user or not user.is_active:
            return {"ok": False, "reason": "Usuario inválido"}

        if not user.password_reset_token or not user.password_reset_expires:
            return {"ok": False, "reason": "Código no solicitado"}

        if datetime.now(timezone.utc) > user.password_reset_expires:
            return {"ok": False, "reason": "Código expirado"}

        if not verify_hash(code, user.password_reset_token):
            return {"ok": False, "reason": "Código inválido"}

        # Hash de la nueva contraseña
        password_hash = hash_value(new_password)

        # Actualizar contraseña y limpiar token
        await self.users.patch_user(
            user.id,
            {
                "password_hash": password_hash,
                "password_reset_token": None,
                "password_reset_expires": None,
            },
        )

        return {"ok": True}

    async def change_password(self, user_id: int, current_password: str, new_password: str):
        """Cambia la contraseña de un usuario autenticado"""
        user = await self.users.find_by_id(user_id)
        if not user or not user.is_active:
            return {"ok": False, "reason": "Usuario inválido"}

        # Verificar que la contraseña actual sea correcta
        if not verify_hash(current_password, user.password_hash):
            return {"ok": False, "reason": "Contraseña actual incorrecta"}

        # Hash de la nueva contraseña
        password_hash = hash_value(new_password)

        # Actualizar contraseña
        await self.users.patch_user(
            user.id,
            {"password_hash": password_hash},
        )

        return {"ok": True}

    async def update_profile(self, user_id: int, profile_data: dict):
        """
        Actualiza el perfil del usuario autenticado.
        Puede actualizar username (users) y datos personales (patients).
        No permite actualizar nombres (first_name, last_name).
        """
        user = await self.users.find_by_id(user_id)
        if not user or not user.is_active:
            raise HTTPException(status_code=401, detail="Usuario inválido")

        # Actualizar username si se proporciona
        if "username" in profile_data and profile_data["username"] is not None:
            # Verificar si el nuevo username ya existe
            existing_username = await self.users.find_by_username(profile_data["username"])
            if existing_username and existing_username.id != user_id:
                raise HTTPException(status_code=409, detail="El username ya está en uso")
            
            await self.users.patch_user(user_id, {"username": profile_data["username"]})

        # Si el usuario es paciente, actualizar datos del paciente
        if user.patient:
            patient_update = {}
            
            # Mapear campos de perfil a campos de paciente
            patient_fields = [
                "phone", "email", "address", "gender", "marital_status",
                "occupation", "education_level", "emergency_contact_name",
                "emergency_contact_relationship", "emergency_contact_phone"
            ]
            
            for field in patient_fields:
                if field in profile_data:
                    patient_update[field] = profile_data[field]
            
            # Si hay campos de paciente para actualizar
            if patient_update:
                from app.db.repositories.patients_repo import PatientsRepo
                patients_repo = PatientsRepo(self.users.db)
                await patients_repo.update(user.patient.id, patient_update)

        # Retornar el usuario actualizado
        return await self.users.find_by_id(user_id)
